Mathlib/Tactic/FailIfNoProgress.lean

```diff
@@ -4,63 +4,203 @@ Released under Apache 2.0 license as described in the file LICENSE.
+not checked, and implementation details are ignored. Expressions are by default compared with
+`isDefEq` at reducible transparency. Any change in the order or number of goals counts as progress.
 
-/-- Run `tacs : TacticM Unit` on `goal`, and fail if no progress is made. -/
-def runAndFailIfNoProgress (goal : MVarId) (tacs : TacticM Unit) : TacticM (List MVarId) := do
+The config preset `FailIfNoProgress.Config.anyChanges` can be used to detect any change;
+`FailIfNoProgress.Config.onlyExprs` will only compare expressions in the target and local context.
+
+For more information, see the documentation for `FailIfNoProgress.Config`.
+-/
+def runAndFailIfNoProgress (goal : MVarId) (tacs : TacticM Unit)
+    (cfg : FailIfNoProgress.Config := {}) : TacticM (List MVarId) := do
+  let decl ← goal.getDecl
   let l ← run goal tacs
-  try
-    let [newGoal] := l | failure
-    guard <|← withNewMCtxDepth <| withReducible <| isDefEq (← newGoal.getType) (← goal.getType)
-    let ctxDecls := (← goal.getDecl).lctx.decls.toList
-    let newCtxDecls := (← newGoal.getDecl).lctx.decls.toList
-    guard <|← withNewMCtxDepth <| withReducible <| lctxIsDefEq ctxDecls newCtxDecls
+  try -- failing this `try` means we return `l`, and occurs iff progress has been made
+    if let .quick := cfg.mode then
+      if ← goal.isAssigned then failure -- progress
+    else
+      let [newGoal] := l | failure -- progress
+      guard <|← goal.compare newGoal cfg.toMVarIdComparisonConfig
+        cfg.toExprComparisonConfig decl
   catch _ =>
     return l
-  throwError "no progress made on {goal}"
+  let resultMsg ← match cfg.mode with
+  | .quick => pure m!"no goals were assigned"
+  | .normal => do pure m!"obtained\n{← getGoals}"
+  throwError "no progress made on goal; {resultMsg}"
+
+/-- Run `tacs`, and fail if no progress is made. Return the result otherwise. By default, this
+compares each of their types and local contexts before and after `tacs` is run; if any changes are
+seen, "progress" has been made, and the tactics succeed. Otherwise, we fail.
+
+`cfg` can be used to specify what kind of comparison to perform when checking for "progress". By
+default, only "observable" changes are checked. For instance, internal `MVarId`s and `FVarId`s are
+not checked, and implementation details are ignored. Expressions are by default compared with
+`isDefEq` at reducible transparency. Any change in the order or number of goals counts as progress.
+
+The config preset `FailIfNoProgress.Config.anyChanges` can be used to detect any change;
+`FailIfNoProgress.Config.onlyExprs` will only compare expressions in the target and local context.
+
+For more information, see the documentation for `FailIfNoProgress.Config`.
+-/
+def failIfNoProgress (tacs : TacticM α) (cfg : FailIfNoProgress.Config := {}) :
+    TacticM α := do
+  let goals₁ ← getGoals
+  let decls₁ ← goals₁.mapM (·.getDecl)
+  let result ← tacs
+  let progress ← match cfg.mode with
+    | .quick => goals₁.anyM (·.isAssigned)
```

Comment:
Hmm, I was curious, and it seems both conventions are used in the case of `let a ← match b with`. My reasoning was that the indentation of the whole term following `←` is one tab greater than that of the `let`. (E.g. if a function's arguments spilled over to the next line, we'd indent them by one tab.) So, by indenting one tab, I'm keeping the `|`'s aligned with the `match`'s indentation (instead of the `let`s). Likewise, consider what happens when `a` becomes longer:
```
let x ← match a with
  | _ => ...

let x ←
  match aaaaaaaaaaaaaaa with
  | _ => ...
```
But maybe that's not how it should work! I'm open to changing it. :)

Reply:
You might be right, and in any case there isn't really a convention written down yet.